mobilitypath_publisher/src/mobilitypath_publisher.cpp

```diff
@@ -0,0 +1,138 @@
+
+    cav_msgs::MobilityPath MobilityPathPublication::mobilityPathMessageGenerator(cav_msgs::TrajectoryPlan trajectory_plan)
+    {
+        cav_msgs::MobilityPath mobility_path_msg;
+        mobility_path_msg.header = composeMobilityHeader();
+        
+
+        std::vector<cav_msgs::TrajectoryPlanPoint> traj_points;
+        traj_points = trajectory_plan.trajectory_points;
+        cav_msgs::Trajectory mob_path_traj = TrajectoryPlantoTrajectory(traj_points);
+        mobility_path_msg.trajectory = mob_path_traj;
+
+        return mobility_path_msg;
+    }
+
+    cav_msgs::MobilityHeader MobilityPathPublication::composeMobilityHeader(){
+        cav_msgs::MobilityHeader header;
+        header.sender_id = sender_id;
+        header.recipient_id = recipient_id;
+        header.sender_bsm_id = sender_bsm_id;
+        // random GUID that identifies this particular plan for future reference
+        header.plan_id = boost::uuids::to_string(boost::uuids::random_generator()());
+        header.timestamp = ros::Time::now().toNSec()/1000; //time in millisecond
+        
+        return header;
+    }
+
+    cav_msgs::Trajectory MobilityPathPublication::TrajectoryPlantoTrajectory(std::vector<cav_msgs::TrajectoryPlanPoint> traj_points){
+        cav_msgs::Trajectory traj;
+        cav_msgs::LocationECEF ecef_location = TrajectoryPointtoECEF(traj_points[0]);
+
+        if (traj_points.size()<2){
+            ROS_WARN("Received Trajectory Plan is too small");
+            traj.offsets = {};
+        }
+        else{
+            for (int i=1; i<traj_points.size(); i++){
+                if (trajectoryPoseCheck(traj_points[i])){
+                    cav_msgs::LocationOffsetECEF offset;
+                    cav_msgs::LocationECEF new_point = TrajectoryPointtoECEF(traj_points[i]);
+                    offset.offset_x = new_point.ecef_x - ecef_location.ecef_x;
+                    offset.offset_y = new_point.ecef_y - ecef_location.ecef_y;
+                    offset.offset_z = new_point.ecef_z - ecef_location.ecef_z;
+                    traj.offsets.push_back(offset);
+                }
+                else{
+                    ecef_location = TrajectoryPointtoECEF(traj_points[i]);
+                    traj.offsets = {};
+                }
+            }
+        }
+        
+        
+        traj.location = ecef_location;
+        return traj;
+    }
+
+    bool MobilityPathPublication::trajectoryPoseCheck(cav_msgs::TrajectoryPlanPoint point){
+        if (fabs(current_pose_->pose.position.x - point.x) > x_threshold || fabs(current_pose_->pose.position.y - point.y) > y_threshold)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    cav_msgs::LocationECEF MobilityPathPublication::TrajectoryPointtoECEF(cav_msgs::TrajectoryPlanPoint traj_point){
+        cav_msgs::LocationECEF ecef_point;    
+        try
+        {
+            geometry_msgs::TransformStamped tf = tf2_buffer_.lookupTransform("earth", "map", ros::Time(0));
```

Comment:
I would not recommend doing this lookup in a loop as it will be pretty slow and earth->map should not be allowed to change between points on a single mobility path message. Perform the lookup just once at the top of your callback before you begin processing the path and pass the transform into this function.

Reply:
Modified